### PR TITLE
IDVA6-1798: Created POC.

### PR DESCRIPTION
**Pull request by krishna-patel-ch**
__Short description outlining key changes/additions__

Created POC

__JIRA Ticket Number__

IDVA6-1798

### Type of change

* [ ] Bug fix
* [x] New feature
* [ ] Breaking change
* [x] Infrastructure change

### Pull request checklist

* [x] I have added unit tests for new code that I have added
* [x] I have added/updated functional tests where appropriate
* [x] The code follows our [coding standards](https://github.com/companieshouse/styleguides/blob/master/java.md)

__An exhaustive list of peer review checks can be read [here](https://github.com/companieshouse/styleguides/blob/master/java_review.md#developer-actions-prior-to-code-commitreview-started)__

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/accounts/association/repositories/AssociationsRepository.java [hunk 1]
```diff
@@ -47,4 +47,10 @@ public interface AssociationsRepository extends MongoRepository<AssociationDao,
     @Query( "{ '$or': [ { 'user_id': { '$ne': null, '$eq': ?0 } }, { 'user_email': { '$ne': null, '$eq': ?1 } } ], 'status': 'awaiting-approval', 'approval_expiry_at': { $gt: ?2 } }" )
     Stream<AssociationDao> fetchAssociationsWithActiveInvitations( final String userId, final String userEmail, final LocalDateTime now );
 
+    @Query( "{ 'status': 'migrated', '$or': [ { 'user_id': null }, { 'user_id': { '$exists': false } } ] }" )
+    Page<AssociationDao> fetchUnprocessedMigratedAssociations( final Pageable pageable );
+
+    @Query( value = "{ 'status': 'migrated', '$or': [ { 'user_id': null }, { 'user_id': { '$exists': false } } ] }", count = true )
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/repositories/AssociationsRepository.java:53]: this should be "and" i think. migrated "and" user_email not null, or user id exists.

File: src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    public DataMigrationPostProcessingTask( final AssociationsService associationsService, final UsersService usersService ) {
+        this.associationsService = associationsService;
+        this.usersService = usersService;
+    }
+
+    // TODO: change cron = "0 0 0 * * ?", so that the cron job runs everyday at midnight
+    @Scheduled( cron = "0 2 17 * * ?" )
+    public void processMigratedAssociations(){
+        LOG.infoContext( DATA_MIGRATION_TASK, "Starting up data migration task...", null );
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:39]: data migration user id update task something.

File: src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+        /* Decided to use pagination because loading in all of the associations at the same time might consume too much memory.
+
+           Updating the associations on a given page will distort the pagination for the subsequent pages, but has no effect on the
+           previous pages. So, decided to start from the last page and work back to the first page, to avoid the distortion effect.
+
+           Suggestion: Reactive programming would be well suited to this task.
+        */
+        final var totalAssociations = associationsService.fetchNumberOfUnprocessedMigratedAssociations();
+        final var totalPages = ( totalAssociations / ITEMS_PER_PAGE ) + ( totalAssociations % 2 == 0 ? 0 : 1 );
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:50]: why do we need this. We will be doing a PUT to update email right.

Comment (kkonuganti-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:50]: or call the task which would use put

File: src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+               Note, that this means that ITEMS_PER_PAGE not only determines batch size, but also the number of
+               users we request from the accounts-user-api in a single call.
+            */
+            final var userEmails = page
+                    .map( AssociationDao::getUserEmail )
+                    .filter( Objects::nonNull )
+                    .toList();
+
+            final var userEmailToIdMapper = new HashMap<String, String>();
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:70]: this will hold the user ids which only have emails.

File: src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            Optional.ofNullable( usersService.searchUserDetails( userEmails ) )
+                    .orElse( new UsersList() )
+                    .forEach( user -> userEmailToIdMapper.put( user.getEmail(), user.getUserId() ) );
+
+            /* This block filters out the associations on the page, so that we skip any associations relating to
+               users that were not found by the accounts-user-api (i.e. users without a CHS account).
+               The remaining associations are processed.
+            */
+            page.filter( associationDao -> userEmailToIdMapper.containsKey( associationDao.getUserEmail() ) )
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:79]: you dont have to filter here... you can straight away find the email id and replace with user id. Since all the emails ids in the mapper above are for users who do not have user id in first place.
We dont care about the users who still have user emails. If the are not found we dont update them.

File: src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        LOG.infoContext(DATA_MIGRATION_USER_ID_UPDATE_TASK, "Starting up data migration user_id task...", null );
+
+        final var totalAssociations = associationsService.fetchNumberOfUnprocessedMigratedAssociations();
+        final var totalPages = ( totalAssociations / ITEMS_PER_PAGE ) + ( totalAssociations % 2 == 0 ? 0 : 1 );
+        int currentPage = (int) totalPages - 1;
+
+        LOG.debugContext(DATA_MIGRATION_USER_ID_UPDATE_TASK, String.format( "Identified %s associations to be processed in %s batches in reverse order... starting with batch %s", totalAssociations, totalPages, currentPage ), null );
+
+        while ( currentPage > -1 ) {
+            final var page = associationsService.fetchUnprocessedMigratedAssociations( currentPage, ITEMS_PER_PAGE );
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:45]: var page = his already gets all the associations where email_id exists.

File: src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+            final var users = usersService.searchUserDetails( page.stream() );
+
+            page.stream()
+                    .filter( associationDao -> users.containsKey( associationDao.getUserEmail() ) )
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:50]: if page already has associations where email exists. why are we filtering again?

File: src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+                    .collect( Collectors.toMap( AssociationDao::getId, associationDao -> users.get( associationDao.getUserEmail() ).getUserId() ) )
+                    .forEach( ( associationId, userId ) -> {
+                        final var update = new Update()
+                                .set( "user_id", userId )
+                                .set( "user_email", null );
+                        associationsService.updateAssociation( associationId, update );
+                        LOG.debugContext(DATA_MIGRATION_USER_ID_UPDATE_TASK, String.format( "Successfully completed migration of association %s", associationId ), null );
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:57]: lets make this info.

File: src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                    } );
+
+            LOG.debugContext(DATA_MIGRATION_USER_ID_UPDATE_TASK, String.format( "Successfully processed batch %s", currentPage ), null );
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:60]: info

File: src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            currentPage--;
+        }
+
+        LOG.infoContext(DATA_MIGRATION_USER_ID_UPDATE_TASK, "Shutting data migration user_id task...", null );
+    }
+
+
+
+
+    /*
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:71]: remove comments

File: src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    private final AssociationsService associationsService;
+
+    private final UsersService usersService;
+
+    private static final Logger LOG = LoggerFactory.getLogger( StaticPropertyUtil.APPLICATION_NAMESPACE );
+
+    private static final String DATA_MIGRATION_USER_ID_UPDATE_TASK = "data-migration-user-id-update-task";
+
+    // TODO: choose different value e.g. 15
+    private static final int ITEMS_PER_PAGE = 2;
```
Comment (kkonuganti-ch) [inline src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:26]: add this to properties

File: src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java (hunk 6 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:45]: I'm not sure why we should remove `final`? The variable never gets updated.

The call to `fetchUnprocessedMigratedAssociations` will only get `ITEMS_PER_PAGE` number of associations on a given iteration e.g. if we set it to 100, it will fetch 100 migrating associations. Then on the next iteration it will fetch the next 100 etc. This lets us control how many are loaded into memory at the same time.

File: src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java (hunk 7 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:50]: You are correct - the page only contains associations where the email exists.

On line 47, we extract the `user_email`s from the associations, and pass it to the accounts-user-api to fetch the users (because we need the 'user_id'). Because the user might, or might not have a CHS account, the accounts-user-api might or might not actually find all of the users. So the `users` variable is a subset of the users from across the associations, i.e. all of the ones that actually have a `user_id`.

On line 50, the filter extracts the 'user_email' from the association, and checks if it exists in 'users' - if it does then this means the user in the association has a 'user_id'.

In other words, this is not the same filter as the one in the query.

File: src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java (hunk 8 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:57]: Updated

File: src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java (hunk 9 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:60]: Updated

File: src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java (hunk 10 above)
Comment (krishna-patel-ch) [reply src/main/java/uk/gov/companieshouse/accounts/association/tasks/DataMigrationPostProcessingTask.java:71]: Removed

**Review by kkonuganti-ch — DISMISSED** [review]

**Comment by ch-code-analysis** [comment]
CI: No security warnings found
